***Работа № 2.***

***1) Задание: загрузить данные из файла (приложены к ниже):*** 

***sp500hst.txt*** 

In [1]:
import pandas as pd

In [2]:
%config IPCompleter.greedy =True

In [3]:
df = pd.read_csv('sp500hst.txt', names=['date', 'name', 'open', 'high', 'low', 'close', 'volume'])
df.head()

date name   open     high     low  close  volume
0  20090821    A  25.60  25.6100  25.220  25.55   34758
1  20090824    A  25.64  25.7400  25.330  25.50   22247
2  20090825    A  25.50  25.7000  25.225  25.34   30891
3  20090826    A  25.32  25.6425  25.145  25.48   33334
4  20090827    A  25.50  25.5700  25.230  25.54   70176

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 122574 entries, 0 to 122573
Data columns (total 7 columns):
date      122574 non-null int64
name      122574 non-null object
open      122574 non-null float64
high      122574 non-null float64
low       122574 non-null float64
close     122574 non-null float64
volume    122574 non-null int64
dtypes: float64(4), int64(2), object(1)
memory usage: 6.5+ MB


In [6]:
df.date = df.date.astype(str)

In [7]:
def get_year(row):
        return row['date'][:4]
    
def get_month(row):
        return row['date'][4:6]
    
def get_day(row):
        return row['date'][6:]

In [8]:
df['year'] = df.apply(get_year, axis=1)
df['month'] = df.apply(get_month, axis=1)
df['day'] = df.apply(get_day, axis=1)

In [9]:
df.drop('date', axis=1, inplace=True)

In [10]:
df.head()

name   open     high     low  close  volume  year month day
0    A  25.60  25.6100  25.220  25.55   34758  2009    08  21
1    A  25.64  25.7400  25.330  25.50   22247  2009    08  24
2    A  25.50  25.7000  25.225  25.34   30891  2009    08  25
3    A  25.32  25.6425  25.145  25.48   33334  2009    08  26
4    A  25.50  25.5700  25.230  25.54   70176  2009    08  27

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 122574 entries, 0 to 122573
Data columns (total 9 columns):
name      122574 non-null object
open      122574 non-null float64
high      122574 non-null float64
low       122574 non-null float64
close     122574 non-null float64
volume    122574 non-null int64
year      122574 non-null object
month     122574 non-null object
day       122574 non-null object
dtypes: float64(4), int64(1), object(4)
memory usage: 8.4+ MB


***3.1) Создать столбец со значениями средними, между значением дневного минимума и максимума.***

In [12]:
df['high_low'] = df['high'] - df['low']

In [13]:
df.describe().loc['mean', 'high_low']

1.04777972245339

In [14]:
df.drop('high_low', axis=1, inplace=True)

***3.2) Рассчитать среднее значение показателей для каждого из столбцов 3-6.***

In [15]:
df.describe().loc['mean', :]

open         42.595458
high         43.102243
low          42.054464
close        42.601865
volume    81395.068138
Name: mean, dtype: float64

***4.1) Создать DataFrame состоящий из строк относящихся только к 13 августа 2010 года.***

In [16]:
df4_1 = df[(df.year == '2010') & (df.month == '08') & (df.day == '13')]

***4.1.2) Из DataFrame, созданного в 4.1, cоздать серию из 30 первых имен***

In [21]:
first_30_series = pd.Series(df4_1.name.unique()[:30])
first_30_series

0        A
1       AA
2     AAPL
3      ABC
4      ABT
5      ACE
6     ADBE
7      ADI
8      ADM
9      ADP
10    ADSK
11     AEE
12     AEP
13     AES
14     AET
15     AFL
16     AGN
17     AIG
18     AIV
19     AIZ
20    AKAM
21     AKS
22     ALL
23    ALTR
24    AMAT
25     AMD
26    AMGN
27     AMP
28     AMT
29    AMZN
dtype: object

***4.2) Из DataFrame, созданного в 4.1, создать новый DataFrame индексом в котором являются имена тикеров.***

In [23]:
df4_2 = df4_1.set_index('name')
df4_2.head()

open    high      low   close  volume  year month day
name                                                         
A      27.39   27.85   27.300   27.35   24469  2010    08  13
AA     10.70   10.79   10.575   10.64  154397  2010    08  13
AAPL  251.03  251.88  249.090  249.10  126738  2010    08  13
ABC    29.01   29.13   28.690   29.04   16196  2010    08  13
ABT    50.13   50.26   49.910   50.06   50879  2010    08  13

***4.3) Создать DataFrame состоящий из строк относящихся только тикеру ALTR.***

In [35]:
df_altr = df[df.name == 'ALTR']
df_altr

name   open   high    low  close  volume  year month day
5544  ALTR  19.36  19.50  19.15  19.48   54580  2009    08  21
5545  ALTR  19.51  19.72  19.04  19.13   42059  2009    08  24
5546  ALTR  19.13  19.39  18.95  19.05   48539  2009    08  25
5547  ALTR  19.11  19.41  18.99  19.28   42058  2009    08  26
5548  ALTR  19.23  19.32  18.88  19.08   63271  2009    08  27
...    ...    ...    ...    ...    ...     ...   ...   ...  ..
5782  ALTR  25.30  25.50  24.94  24.94   69416  2010    08  13
5783  ALTR  24.82  25.31  24.53  25.11   60276  2010    08  16
5784  ALTR  25.22  25.88  25.04  25.54   89329  2010    08  17
5785  ALTR  25.61  25.96  25.24  25.50   61815  2010    08  19
5786  ALTR  25.44  25.83  25.32  25.60   54383  2010    08  20

[243 rows x 9 columns]

***4.4) Из DataFrame, созданного в 4.3, создать DataFrame индексом в котором являются даты.***

In [112]:
df_altr_date_index = df_altr.set_index(['year', 'month', 'day'])
print(df_altr_date_index.loc['2009'])
df_altr_date_index

           name   open   high    low  close  volume
month day                                          
08    21   ALTR  19.36  19.50  19.15  19.48   54580
      24   ALTR  19.51  19.72  19.04  19.13   42059
      25   ALTR  19.13  19.39  18.95  19.05   48539
      26   ALTR  19.11  19.41  18.99  19.28   42058
      27   ALTR  19.23  19.32  18.88  19.08   63271
...         ...    ...    ...    ...    ...     ...
12    24   ALTR  22.88  23.18  22.82  23.14   12363
      28   ALTR  23.09  23.14  22.65  22.83   27074
      29   ALTR  22.80  22.86  22.41  22.57   36495
      30   ALTR  22.51  22.91  22.41  22.81   21256
      31   ALTR  22.79  22.93  22.61  22.63   30705

[85 rows x 6 columns]


name   open   high    low  close  volume
year month day                                          
2009 08    21   ALTR  19.36  19.50  19.15  19.48   54580
           24   ALTR  19.51  19.72  19.04  19.13   42059
           25   ALTR  19.13  19.39  18.95  19.05   48539
           26   ALTR  19.11  19.41  18.99  19.28   42058
           27   ALTR  19.23  19.32  18.88  19.08   63271
...              ...    ...    ...    ...    ...     ...
2010 08    13   ALTR  25.30  25.50  24.94  24.94   69416
           16   ALTR  24.82  25.31  24.53  25.11   60276
           17   ALTR  25.22  25.88  25.04  25.54   89329
           19   ALTR  25.61  25.96  25.24  25.50   61815
           20   ALTR  25.44  25.83  25.32  25.60   54383

[243 rows x 6 columns]

***4.5) В DataFrame, созданном в 4.3, добавить столбец, содержащий только число месяца, к которому относится дата.*** 

In [34]:
df_altr.month

5544    08
5545    08
5546    08
5547    08
5548    08
        ..
5782    08
5783    08
5784    08
5785    08
5786    08
Name: month, Length: 243, dtype: object

***4.5.1) Создать DataFrame состоящий из строк относящихся только тикеру WAT и объединить его с DataFrame для ALTR.*** 

In [36]:
df_wat = df[df.name == 'WAT']
df_altr_wat = pd.concat([df_altr, df_wat], axis=0)
df_altr_wat

name   open   high    low  close  volume  year month day
5544    ALTR  19.36  19.50  19.15  19.48   54580  2009    08  21
5545    ALTR  19.51  19.72  19.04  19.13   42059  2009    08  24
5546    ALTR  19.13  19.39  18.95  19.05   48539  2009    08  25
5547    ALTR  19.11  19.41  18.99  19.28   42058  2009    08  26
5548    ALTR  19.23  19.32  18.88  19.08   63271  2009    08  27
...      ...    ...    ...    ...    ...     ...   ...   ...  ..
115486   WAT  62.84  63.32  62.35  62.39   11458  2010    08  13
115487   WAT  62.00  62.36  61.46  61.95   12828  2010    08  16
115488   WAT  62.74  64.79  62.56  64.43    9557  2010    08  17
115489   WAT  64.10  64.38  62.43  62.53    8350  2010    08  19
115490   WAT  61.99  62.73  61.44  62.68    9270  2010    08  20

[488 rows x 9 columns]

***4.5.2) Создать DataFrame в котором присутствует столбец, отражающий разницу в объемах торгов по WAT и ALTR в одинаковые дни и содержит исходные данные об объеме торгов этими акциями.***

In [37]:
df_altr.shape, df_wat.shape

((243, 9), (245, 9))

In [43]:
df_diff_altr_wat = df_altr.merge(df_wat, how='inner', on=['year', 'month', 'day'])
df_diff_altr_wat

name_x  open_x  high_x  low_x  close_x  volume_x  year month day name_y  \
0     ALTR   19.36   19.50  19.15    19.48     54580  2009    08  21    WAT   
1     ALTR   19.51   19.72  19.04    19.13     42059  2009    08  24    WAT   
2     ALTR   19.13   19.39  18.95    19.05     48539  2009    08  25    WAT   
3     ALTR   19.11   19.41  18.99    19.28     42058  2009    08  26    WAT   
4     ALTR   19.23   19.32  18.88    19.08     63271  2009    08  27    WAT   
..     ...     ...     ...    ...      ...       ...   ...   ...  ..    ...   
238   ALTR   25.30   25.50  24.94    24.94     69416  2010    08  13    WAT   
239   ALTR   24.82   25.31  24.53    25.11     60276  2010    08  16    WAT   
240   ALTR   25.22   25.88  25.04    25.54     89329  2010    08  17    WAT   
241   ALTR   25.61   25.96  25.24    25.50     61815  2010    08  19    WAT   
242   ALTR   25.44   25.83  25.32    25.60     54383  2010    08  20    WAT   

     open_y  high_y  low_y  close_y  volume_y  
0     50.77  52.144  50.33    51.80      9785  
1     52.02  52.020  51.38    51.48      5866  
2     51.61  52.740  51.50    52.23      9567  
3     51.94  52.860  51.93    52.17      4202  
4     52.09  52.220  51.33    51.73      6473  
..      ...     ...    ...      ...       ...  
238   62.84  63.320  62.35    62.39     11458  
239   62.00  62.360  61.46    61.95     12828  
240   62.74  64.790  62.56    64.43      9557  
241   64.10  64.380  62.43    62.53      8350  
242   61.99  62.730  61.44    62.68      9270  

[243 rows x 15 columns]

In [45]:
df_diff_altr_wat['volume_difference'] = df_diff_altr_wat.volume_x - df_diff_altr_wat.volume_y
df_diff_altr_wat

name_x  open_x  high_x  low_x  close_x  volume_x  year month day name_y  \
0     ALTR   19.36   19.50  19.15    19.48     54580  2009    08  21    WAT   
1     ALTR   19.51   19.72  19.04    19.13     42059  2009    08  24    WAT   
2     ALTR   19.13   19.39  18.95    19.05     48539  2009    08  25    WAT   
3     ALTR   19.11   19.41  18.99    19.28     42058  2009    08  26    WAT   
4     ALTR   19.23   19.32  18.88    19.08     63271  2009    08  27    WAT   
..     ...     ...     ...    ...      ...       ...   ...   ...  ..    ...   
238   ALTR   25.30   25.50  24.94    24.94     69416  2010    08  13    WAT   
239   ALTR   24.82   25.31  24.53    25.11     60276  2010    08  16    WAT   
240   ALTR   25.22   25.88  25.04    25.54     89329  2010    08  17    WAT   
241   ALTR   25.61   25.96  25.24    25.50     61815  2010    08  19    WAT   
242   ALTR   25.44   25.83  25.32    25.60     54383  2010    08  20    WAT   

     open_y  high_y  low_y  close_y  volume_y  volume_difference  
0     50.77  52.144  50.33    51.80      9785              44795  
1     52.02  52.020  51.38    51.48      5866              36193  
2     51.61  52.740  51.50    52.23      9567              38972  
3     51.94  52.860  51.93    52.17      4202              37856  
4     52.09  52.220  51.33    51.73      6473              56798  
..      ...     ...    ...      ...       ...                ...  
238   62.84  63.320  62.35    62.39     11458              57958  
239   62.00  62.360  61.46    61.95     12828              47448  
240   62.74  64.790  62.56    64.43      9557              79772  
241   64.10  64.380  62.43    62.53      8350              53465  
242   61.99  62.730  61.44    62.68      9270              45113  

[243 rows x 16 columns]

***4.6) В DataFrame добавить столбец, содержащий только число месяца, к которому относится дата.***

In [46]:
df.day

0         21
1         24
2         25
3         26
4         27
          ..
122569    13
122570    16
122571    17
122572    19
122573    20
Name: day, Length: 122574, dtype: object

***4.7) Из DataFrame выбрать только строки, относящиеся к 13 числу месяца.***

In [47]:
df[df.day=='13']

name   open   high     low  close  volume  year month day
35        A  27.41  27.53  27.150  27.29   14226  2009    10  13
53        A  27.97  28.89  27.750  28.61   62633  2009    11  13
93        A  30.47  30.78  30.050  30.69   24455  2010    01  13
154       A  34.15  34.23  33.880  33.90   25868  2010    04  13
176       A  34.83  35.29  34.430  34.78   43992  2010    05  13
...     ...    ...    ...     ...    ...     ...   ...   ...  ..
122422  ZMH  62.26  62.79  61.732  62.55    7663  2010    01  13
122483  ZMH  59.27  60.15  59.060  60.09   13177  2010    04  13
122505  ZMH  60.97  61.20  60.650  60.65   10647  2010    05  13
122546  ZMH  56.38  57.57  56.380  56.92   21035  2010    07  13
122569  ZMH  51.72  51.90  51.380  51.44   14561  2010    08  13

[3510 rows x 9 columns]

***4) Выявить столбцы, содержащие минимальные и максимальные значения (если такие столбцы есть). (Проверить, есть ли среди столбцов, столбец с минимальным значением курса акций.)***

Если честно, не очень понятен вопрос

In [101]:
import numpy as np

In [50]:
df.describe().loc[['min', 'max'], :]

open    high     low   close      volume
min    1.36    1.39    1.00    1.35        47.0
max  627.84  629.51  624.24  626.75  21474836.0

In [100]:
df.loc[[np.where(df.open == df.open.min())[0][0], 
       np.where(df.open == df.open.max())[0][0],
       np.where(df.high == df.high.min())[0][0],
       np.where(df.high == df.high.max())[0][0],
       np.where(df.low == df.low.min())[0][0],
       np.where(df.low == df.low.max())[0][0],
       np.where(df.close == df.close.min())[0][0],
       np.where(df.close == df.close.max())[0][0],
       np.where(df.volume == df.volume.min())[0][0], 
       np.where(df.volume == df.volume.max())[0][0]], :]

name    open    high     low   close    volume  year month day
39052  ETFC    1.36    1.42    1.34    1.40    592350  2009    08  26
48036  GOOG  627.84  627.84  621.54  623.99     30078  2010    01  05
39051  ETFC    1.38    1.39    1.34    1.35    266104  2009    08  25
48035  GOOG  627.46  629.51  624.24  626.75     19579  2010    01  04
24195   CNP   14.39   14.39    1.00   13.88     93886  2010    05  06
48035  GOOG  627.46  629.51  624.24  626.75     19579  2010    01  04
39051  ETFC    1.38    1.39    1.34    1.35    266104  2009    08  25
48035  GOOG  627.46  629.51  624.24  626.75     19579  2010    01  04
96524    RX   21.97   21.98   21.96   21.98        47  2010    02  26
17114     C    3.26    3.40    3.21    3.40  21474836  2009    12  18

Столбец 24195 имеет минимальное значение курса акций

***5) По данным из 'sp500hst.txt' при помощи Pandas рассчитать максимальные значения в 3-6 столбца для дат после 20100801***

In [137]:
dates = ['year', 'month', 'day']
df2 = df.copy()
for i in dates:
    df2[i] = df2[i].astype('int')

In [138]:
df2.describe()

open           high            low          close  \
count  122574.000000  122574.000000  122574.000000  122574.000000   
mean       42.595458      43.102243      42.054464      42.601865   
std        42.909787      43.378185      42.397805      42.900199   
min         1.360000       1.390000       1.000000       1.350000   
25%        21.810000      22.100000      21.500000      21.810000   
50%        35.130000      35.540000      34.690000      35.130000   
75%        51.840000      52.450000      51.230000      51.870000   
max       627.840000     629.510000     624.240000     626.750000   

             volume          year          month            day  
count  1.225740e+05  122574.00000  122574.000000  122574.000000  
mean   8.139507e+04    2009.64818       6.487640      15.694519  
std    3.423884e+05       0.47754       3.417996       8.624995  
min    4.700000e+01    2009.00000       1.000000       1.000000  
25%    1.661150e+04    2009.00000       4.000000       9.000000  
50%    3.303500e+04    2010.00000       6.000000      16.000000  
75%    6.775600e+04    2010.00000       9.000000      23.000000  
max    2.147484e+07    2010.00000      12.000000      31.000000

In [146]:
round(df2[((df2.year >= 2010) & (df2.month >= 8) & (df2.day >= 1))   | (df2.year >= 2011)].describe().loc['max', 'open':'close'], 2).reset_index()

index     max
0   open  505.85
1   high  508.60
2    low  503.56
3  close  508.10

***6) По данным из 'sp500hst.txt' при помощи Pandas рассчитать средний объем торгов (обеъм торгов - в последнем столбце) при значениях 3го столбца больше 50.***

In [148]:
round(df[df.open > 50].mean()['volume'], 2)

36467.09

***7) Добавить столбец с расшифровкой названия тикера используя данные из файла sp_data2.csv . В случае нехватки данных об именах тикеров корректно обработать их.***

In [182]:
tickers = pd.read_csv('sp_data2.csv', sep=';', names = ['name', 'full_name', 'share'])
tickers.head()

name   full_name share
0   AAPL       Apple  3.6%
1   AMZN  Amazon.com  3.2%
2  GOOGL    Alphabet  3.1%
3   GOOG    Alphabet  3.1%
4   MSFT   Microsoft  3.0%

In [183]:
df.name.unique()

array(['A', 'AA', 'AAPL', 'ABC', 'ABT', 'ACE', 'ACS', 'ADBE', 'ADI',
       'ADM', 'ADP', 'ADSK', 'AEE', 'AEP', 'AES', 'AET', 'AFL', 'AGN',
       'AIG', 'AIV', 'AIZ', 'AKAM', 'AKS', 'ALL', 'ALTR', 'AMAT', 'AMD',
       'AMGN', 'AMP', 'AMT', 'AMZN', 'AN', 'ANF', 'AOC', 'AON', 'APA',
       'APC', 'APD', 'APH', 'APOL', 'ARG', 'ATI', 'AVB', 'AVP', 'AVY',
       'AXP', 'AYE', 'AZO', 'BA', 'BAC', 'BAX', 'BBBY', 'BBT', 'BBY',
       'BCR', 'BDK', 'BDX', 'BEN', 'BF.B', 'BHI', 'BIG', 'BIIB', 'BJS',
       'BK', 'BLL', 'BMC', 'BMS', 'BMY', 'BNI', 'BRCM', 'BRK.B', 'BSX',
       'BTU', 'BXP', 'C', 'CA', 'CAG', 'CAH', 'CAM', 'CAT', 'CB', 'CBE',
       'CBG', 'CBS', 'CCE', 'CCL', 'CEG', 'CELG', 'CEPH', 'CERN', 'CF',
       'CFN', 'CHK', 'CHRW', 'CI', 'CIEN', 'CINF', 'CL', 'CLF', 'CLX',
       'CMA', 'CMCSA', 'CME', 'CMI', 'CMS', 'CNP', 'CNX', 'COF', 'COG',
       'COH', 'COL', 'COP', 'COST', 'CPB', 'CPWR', 'CRM', 'CSC', 'CSCO',
       'CSX', 'CTAS', 'CTL', 'CTSH', 'CTXS', 'CVG', 'CVH', 'CVS', 'CVX

In [184]:
df3 = df.merge(tickers, how='left', on ='name')

In [186]:
ticker_names_not_found = df3[df3.full_name.fillna('Unknown') == 'Unknown'].name.unique()

In [180]:
#df.drop(['full_name_x', 'full_name_y', 'share_x', 'share_y'], axis=1, inplace=True)

In [188]:
df3.loc[df3.full_name.fillna('Unknown') == 'Unknown', 'share'] = 0
df3.loc[df3.full_name.fillna('Unknown') == 'Unknown', 'full_name'] = df['name']

In [189]:
df3.isnull().sum()

name         0
open         0
high         0
low          0
close        0
volume       0
year         0
month        0
day          0
full_name    0
share        0
dtype: int64

In [191]:
df3[df3.name == 'JAVA']

name   open  high   low  close  volume  year month day full_name share
58343  JAVA  9.350  9.36  9.33   9.35  148943  2009    08  21      JAVA     0
58344  JAVA  9.350  9.36  9.32   9.32   77967  2009    08  24      JAVA     0
58345  JAVA  9.340  9.35  9.32   9.34   51361  2009    08  25      JAVA     0
58346  JAVA  9.330  9.36  9.32   9.35   81984  2009    08  26      JAVA     0
58347  JAVA  9.350  9.36  9.33   9.34   51132  2009    08  27      JAVA     0
...     ...    ...   ...   ...    ...     ...   ...   ...  ..       ...   ...
58439  JAVA  9.410  9.45  9.41   9.43   54027  2010    01  20      JAVA     0
58440  JAVA  9.470  9.48  9.47   9.47  126642  2010    01  21      JAVA     0
58441  JAVA  9.470  9.48  9.46   9.46   90326  2010    01  22      JAVA     0
58442  JAVA  9.475  9.48  9.47   9.48   36967  2010    01  25      JAVA     0
58443  JAVA  9.470  9.49  9.46   9.49  118044  2010    01  26      JAVA     0

[101 rows x 11 columns]

***8) Вывести тикеры, для которых не были найдены расшифровки.***

In [194]:
ticker_names_not_found

array(['AA', 'ACE', 'ACS', 'AKS', 'ALTR', 'AN', 'ANF', 'AOC', 'APOL',
       'ARG', 'ATI', 'AVP', 'AYE', 'BBBY', 'BCR', 'BDK', 'BHI', 'BIG',
       'BJS', 'BMC', 'BMS', 'BNI', 'BRCM', 'BTU', 'CAM', 'CBE', 'CBG',
       'CCE', 'CEG', 'CEPH', 'CFN', 'CHK', 'CIEN', 'CLF', 'CNX', 'COH',
       'CPWR', 'CSC', 'CVG', 'CVH', 'DD', 'DELL', 'DF', 'DNB', 'DNR',
       'DO', 'DOW', 'DTV', 'DV', 'DYN', 'EK', 'EMC', 'EP', 'ERTS', 'ESV',
       'ETFCD', 'FDO', 'FHN', 'FII', 'FO', 'FPL', 'FRX', 'FSLR', 'FTR',
       'GAS', 'GCI', 'GENZ', 'GME', 'GNW', 'GR', 'HAR', 'HCBK', 'HCN',
       'HNZ', 'HOT', 'HSP', 'IGT', 'ITT', 'JAVA', 'JBL', 'JCP', 'JDSU',
       'JNS', 'KBH', 'KFT', 'KG', 'LIFE', 'LLTC', 'LM', 'LO', 'LSI',
       'LTD', 'LUK', 'LXK', 'MBI', 'MDP', 'MEE', 'MFE', 'MHP', 'MHS',
       'MI', 'MIL', 'MJN', 'MOLX', 'MON', 'MOT', 'MTW', 'MUR', 'MWV',
       'MWW', 'NBR', 'NOVL', 'NSM', 'NU', 'NVLS', 'NYT', 'NYX', 'ODP',
       'OI', 'PBG', 'PBI', 'PCL', 'PCLN', 'PCP', 'PCS', 'PDCO', 'PGN',
      

***9) Рассчитать суммарный объем торгов (столбец 7) для одинаковых значений (тикеров) из столбца 2***

In [197]:
df_volume = df.groupby('name').sum()['volume'].reset_index()
df_volume

name    volume
0       A   8609336
1      AA  81898998
2    AAPL  52261170
3     ABC   9006756
4     ABT  18975870
..    ...       ...
519   XTO  21297931
520  YHOO  56837171
521   YUM  10971538
522  ZION  15551119
523   ZMH   4938916

[524 rows x 2 columns]

***9.2) Для каждого тикера подсчитать, сколько раз он встречается.***

In [199]:
df.name.value_counts().reset_index()

index  name
0      SVU   245
1      DNR   245
2      RHT   245
3       EP   245
4      ZMH   245
..     ...   ...
519    SGP    47
520    NEE    41
521    WYE    38
522    QEP    35
523  ETFCD    20

[524 rows x 2 columns]

***10) Рассчитать среднее значение показателей для каждого из столбцов 3-6 для одинаковых значений (тикеров) из столбца 2 и сохранить в новом DataFrame***

In [202]:
df_mean_ticker_value = df.groupby('name').mean()[['open', 'high', 'low', 'close']].reset_index()
df_mean_ticker_value

name        open        high         low       close
0       A   30.234857   30.567838   29.866467   30.246143
1      AA   13.086959   13.275341   12.842954   13.047345
2    AAPL  221.342427  223.556147  218.409411  221.125341
3     ABC   27.432122   27.683660   27.126983   27.429347
4     ABT   50.996776   51.365541   50.555472   50.986490
..    ...         ...         ...         ...         ...
519   XTO   44.640338   45.057131   44.132049   44.588889
520  YHOO   15.854774   16.050593   15.646220   15.847130
521   YUM   37.456898   37.816882   37.094593   37.479633
522  ZION   19.694057   20.171136   19.226905   19.702861
523   ZMH   56.220980   56.799984   55.648306   56.246122

[524 rows x 5 columns]

***11) Добавить в DataFrame новый столбец, содержащий объем торгов (столбец 7), нормированный по среднему значению объема торгов для данного тикера.*** 

In [207]:
#не понял вопрос
df_11 = df.groupby('name').mean()['volume'].reset_index()
df_11

name         volume
0       A   35140.146939
1      AA  334281.624490
2    AAPL  216851.327801
3     ABC   36762.269388
4     ABT   77452.530612
..    ...            ...
519   XTO  102888.555556
520  YHOO  233897.823045
521   YUM   44781.787755
522  ZION   63734.094262
523   ZMH   20158.840816

[524 rows x 2 columns]

In [211]:
df_mean_ticker_value['volume_normed'] = df_11.volume / df_volume.volume

In [214]:
df_mean_ticker_value

name        open        high         low       close  volume_normed
0       A   30.234857   30.567838   29.866467   30.246143       0.004082
1      AA   13.086959   13.275341   12.842954   13.047345       0.004082
2    AAPL  221.342427  223.556147  218.409411  221.125341       0.004149
3     ABC   27.432122   27.683660   27.126983   27.429347       0.004082
4     ABT   50.996776   51.365541   50.555472   50.986490       0.004082
..    ...         ...         ...         ...         ...            ...
519   XTO   44.640338   45.057131   44.132049   44.588889       0.004831
520  YHOO   15.854774   16.050593   15.646220   15.847130       0.004115
521   YUM   37.456898   37.816882   37.094593   37.479633       0.004082
522  ZION   19.694057   20.171136   19.226905   19.702861       0.004098
523   ZMH   56.220980   56.799984   55.648306   56.246122       0.004082

[524 rows x 6 columns]

In [213]:
df_mean_ticker_value.volume_normed.describe()

count    524.000000
mean       0.004710
std        0.003172
min        0.004082
25%        0.004082
50%        0.004082
75%        0.004098
max        0.050000
Name: volume_normed, dtype: float64

In [215]:
#вопрос остался не понятым